### DISCUSSION TITLE
pwpush-ephemeral: How to Add Persistence?

### DISCUSSION THREAD

**Discussion by tschloss** [pglombardo/PasswordPusher#448] [Q&A] [answered]
What is the background of the behavior "in the sqlite3 version of the container the data does not survive container restarts"?

Is it just that is was not important enough so the logic to check for an existing file before creating a new one wasn't implemented yet? Or is it more complex than expected?

I am asking because sqlite would be great for me, but although data not hugely important some level of persistence would be great.

**Comment by pglombardo** [comment, marked as answer]
There are a couple reasons:

1.`pwpush-ephemeral` was intended to throw up a quick instance, share some passwords and then delete all existence that it ever existed
2. sqlite3 doesn't support concurrency so more than 1 user on the same record and errors start happening.  It would be problematic for me to support that officially.

But not all hope is lost.  If you don't want to run the [Postgres or Mysql flavors](https://github.com/pglombardo/PasswordPusher#with-docker-compose), you can make the ephemeral version persist data by providing a Docker bind mount for the sqlite3 file.

For the pwpush-ephemeral container, the path to the sqlite3 file is `/opt/PasswordPusher/db/db.sqlite3`.  

You can't overlay that whole directory because there are other bits in there but you could overlay that file with your own external sqlite3 file.  (even a zero byte file created with `touch` should work - the app should build out the datafile)

    --mount type=bind,source=/path/db.sqlite3,target=/opt/PasswordPusher/db/db.sqlite3

I hope this helps.  If you have any issues, let me know.

**Reply by tschloss** [reply]
Wow - this is truly "actively" maintained 👍 
Thank you, understood the concurrency issue (which would be no issue for me). A Postgres container is almost always running nearby, but I would prefer to be be dependency free (also often PG containers are linked through docker-compose to an app container).

Thanks so far - I will try in the near future!

**Reply by jonnjonzzn** [reply]
This did not work for me but I'll admit to being a container novice but trying.  Container will start but when I visit the site I get:

#500
We're sorry, but something went wrong.

Also noticed the private.sqlite3 file remained seemingly untouched (0 size) after it started.

Command tried:
{docker,podman} run -d -p "8000:5100" --mount type=bind,source=/pods/data/PasswordPusher/private.sqlite3,target=/opt/PasswordPusher/db/private.sqlite3 pglombardo/pwpush-ephemeral:release
I also tried suing docker-compose to same result.

Understand this isn't even supported but it intrigued me as I'm having trouble with name resolution between containers when trying to use the mysql or postgres images.

**Reply by pglombardo** [reply]
Hi @jonnjonzzn,

> Also noticed the private.sqlite3 file remained seemingly untouched (0 size) after it started.

This was also just pointed out by @assices in #1058.  The Docker containers were revamped recently and the name is always `db.sqlite3` now.  See [here](https://github.com/pglombardo/PasswordPusher/blob/master/containers/docker/pwpush-ephemeral/Dockerfile#L18) for the `DATABASE_URL` specification.

I updated the answer above.

> Understand this isn't even supported but it intrigued me as I'm having trouble with name resolution between containers when trying to use the mysql or postgres images.

This one is a bit tricky and I'm not overly knowledgeable about it.  But I do know that when using Docker compose, [this line](https://github.com/pglombardo/PasswordPusher/blob/master/containers/docker/pwpush-mysql/docker-compose-mariadb.yml#L22-L23) works:

```
links:
     - db:mysql
```

So the container can resolve the DNS name `mysql`.

**Comment by pglombardo** [comment]
You are right @jonnjonzzn.  The ephemeral container doesn't automatically setup the database.  I'll fix this soon and publish a new container image.

In the meantime, if you want to fix your container that is giving you the `500` errors:

1. Get the container ID from `docker ps`
2. Run `docker exec -t <container ID> rake db:migrate`

I'll post back once I push out another container with the fix.

**Comment by pglombardo** [comment]
Hi all, with `pwpush-ephemeral` containers `1.19.1` and on, they now run database migrations on container boot.

This should allow you to replace the sqlite3 file [as described in this comment above](https://github.com/pglombardo/PasswordPusher/discussions/448#discussioncomment-3979237)  and have everything work out of the box.

Please let me know if any issues remain.
